modules/header-chain/src/lib.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Defines traits which represent a common interface for Substrate pallets which want to
+//! incorporate bridge functinality.
+
+use frame_support::dispatch::DispatchResult;
+
+/// A base trait for tracking finalized headers on foreign chains in a Substrate pallet. Should be
+/// used as a primitive for more complicated header tracking logic.
+pub trait FinalityHeaderChain<AccountId> {
+	/// The header used by the chain.
+	type Header;
+	/// Any extra data which helps describe a transaction.
+	type Extra;
+	/// The type of transaction proof used by the chain.
+	type Proof;
```

Comment:
The `Proofs` are going to be different for `FullHeaderChain` and `FinalityHeaderChain`. IMHO it's a bit too early to split these two, we don't really know exactly how it would work, so I'd rather defer this split until we have more knowledge.